x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticationService.java

```diff
@@ -690,6 +694,45 @@ void writeAuthToContext(Authentication authentication) {
         private void authenticateToken(AuthenticationToken token) {
             this.consumeToken(token);
         }
+
+        private User maybeMergeAnonymousRolesForUser(User user) {
+            if (SystemUser.is(user) || XPackUser.is(user) || XPackSecurityUser.is(user) || AsyncSearchUser.is(user)) {
+                return user;
+            } else if (isAnonymousUserEnabled && anonymousUser.equals(user) == false) {
+                if (anonymousUser.roles().length == 0) {
+                    throw new IllegalStateException("anonymous is only enabled when the anonymous user has roles");
+                }
+                User userWithMergedRoles = new User(
+                    user.principal(),
+                    mergeAnonymousRoles(user.roles()),
+                    user.fullName(),
+                    user.email(),
+                    user.metadata(),
+                    user.enabled()
+                );
+                if (user.isRunAs()) {
+                    final User authenticatedUserWithMergedRoles = new User(
+                        user.authenticatedUser().principal(),
+                        mergeAnonymousRoles(user.authenticatedUser().roles()),
+                        user.authenticatedUser().fullName(),
+                        user.authenticatedUser().email(),
+                        user.authenticatedUser().metadata(),
+                        user.authenticatedUser().enabled()
+                    );
+                    userWithMergedRoles = new User(userWithMergedRoles, authenticatedUserWithMergedRoles);
+                }
+                return userWithMergedRoles;
+            } else {
+                return user;
+            }
+        }
+
+        private String[] mergeAnonymousRoles(String[] existingRoles) {
+            String[] mergedRoles = new String[existingRoles.length + anonymousUser.roles().length];
+            System.arraycopy(existingRoles, 0, mergedRoles, 0, existingRoles.length);
```

Comment:
Shouldn't we create a Set here and `Collections.addAll` so that we don't end up with duplicate roles in case the anonymous roles contains a role the user already has ?

Reply:
I intentially didn't do deduplication here. Deduplication, filtering out un-resolvable roles, preempting the superuser role are all done in [CompositeRolesStore#getRoles](https://github.com/elastic/elasticsearch/blob/7.6/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java#L197), which comes after authentication.

Even without the anonymous roles, it is still possible to create an user with duplicated roles, e.g. 
`{"roles":["x","x","x"],"password":...}` and these roles will all show up in the authentication response. So I decided to retain the existing behaviour.

Reply:
I don't think we should entertain this existing behavior for no reason, and since we go in the trouble of merging the roles we should do de-duplication here . We could tackle this in a follow up where we for instance should not allow a user to be created with `"roles" :["x","x","x"]` in the first place as this leniency makes no sense and I would argue that if someone created a user with  `"roles" :["x","x"]` , they probably meant  `"roles" :["x","y"]` and mistyped , so it's better to tell them up front/

Reply:
Merging only happens when anonymous access is enabled. If we de-duplicate here, it must also be applied when anonymous access is *not* enabled. So I [re-arranged](https://github.com/elastic/elasticsearch/pull/53453/files#diff-1436647fa05b4a2d5b2712fdefafa98bR702-R741) the code.